kiva/blend2d.py

```diff
@@ -0,0 +1,780 @@
+        """
+        raise NotImplementedError()
+
+    # ----------------------------------------------------------------
+    # Color space manipulation
+    #
+    # I'm not sure we'll mess with these at all.  They seem to
+    # be for setting the color system.  Hard coding to RGB or
+    # RGBA for now sounds like a reasonable solution.
+    # ----------------------------------------------------------------
+
+    def set_fill_color_space(self):
+        msg = "set_fill_color_space not implemented for blend2d yet."
+        raise NotImplementedError(msg)
+
+    def set_stroke_color_space(self):
+        msg = "set_stroke_color_space not implemented for blend2d yet."
+        raise NotImplementedError(msg)
+
+    def set_rendering_intent(self):
+        msg = "set_rendering_intent not implemented for blend2d yet."
+        raise NotImplementedError(msg)
+
+    # ----------------------------------------------------------------
+    # Color manipulation
+    # ----------------------------------------------------------------
+
+    def set_fill_color(self, color):
+        self.gc.set_fill_style(color)
+
+    def set_stroke_color(self, color):
+        self.gc.set_stroke_style(color)
+
+    def set_alpha(self, alpha):
+        self.gc.set_alpha(alpha)
+
+    # ----------------------------------------------------------------
+    # Gradients
+    # ----------------------------------------------------------------
+
+    def linear_gradient(self, x1, y1, x2, y2, stops, spread_method,
+                        units="userSpaceOnUse"):
+        """ Sets a linear gradient as the current brush.
+        """
+        gradient = blend2d.LinearGradient(x1, y1, x2, y2)
+        gradient.extend_mode = gradient_spread_modes.get(
+            spread_method, blend2d.ExtendMode.PAD
+        )
+        for stop in stops:
+            gradient.add_stop(stop[0], stop[1:])
+        self.gc.set_fill_style(gradient)
+
+    def radial_gradient(self, cx, cy, r, fx, fy, stops, spread_method,
+                        units="userSpaceOnUse"):
+        """ Sets a radial gradient as the current brush.
+        """
+        gradient = blend2d.RadialGradient(cx, cy, fx, fy, r)
+        gradient.extend_mode = gradient_spread_modes.get(
+            spread_method, blend2d.ExtendMode.PAD
+        )
+        for stop in stops:
+            gradient.add_stop(stop[0], stop[1:])
+        self.gc.set_fill_style(gradient)
+
+    # ----------------------------------------------------------------
+    # Drawing Images
+    # ----------------------------------------------------------------
+
+    def draw_image(self, img, rect=None):
+        """
+        img is either a N*M*3 or N*M*4 numpy array, or a PIL Image
+
+        rect - a tuple (x, y, w, h)
+        """
+        from PIL import Image
+
+        def normalize_image(img):
+            if not img.mode.startswith("RGB"):
+                img = img.convert("RGB")
+            return img
+
+        if isinstance(img, np.ndarray):
```

Comment:
This code is now more or less repeated across multiple backends - apart from a few extra clauses I think. It might be worth thinking about refactoring them into a single place - maybe put somewhere in `kiva.abstract_graphics_context`.

Reply:
We should open an issue.